src/cc65/loadexpr.c

```diff
@@ -188,34 +200,57 @@ void LoadExpr (unsigned Flags, struct ExprDesc* Expr)
                 Internal ("Invalid location in LoadExpr: 0x%04X", ED_GetLoc (Expr));
         }
 
-        /* Handle bit fields. The actual type may have been casted or
-        ** converted, so be sure to always use unsigned ints for the
-        ** operations.
+        /* Handle bit fields if necessary. The actual type may have been casted or converted,
+        ** so be sure to always use unsigned ints for the operations.
         */
-        if (ED_IsBitField (Expr)) {
+        if (AdjustBitField) {
             /* If the field was loaded as a char, force the shift/mask ops to be char ops.
             ** If it is a char, the load has already put 0 in the upper byte, so that can remain.
             ** CF_FORCECHAR does nothing if the type is not CF_CHAR.
             */
             unsigned F = Flags | CF_FORCECHAR | CF_CONST;
 
-            /* Shift right by the bit offset */
-            g_asr (F, Expr->BitOffs);
-
-            /* Since we have now shifted down, we can do char ops as long as the width fits in
-            ** a char.
+            /* We always need to do something with the low byte, so there is no opportunity
+            ** for optimization by skipping it.
             */
-            if (Expr->BitWidth <= CHAR_BITS) {
-                F |= CF_CHAR;
-            }
+            CHECK (Expr->BitOffs < CHAR_BITS);
 
-            /* And by the width if the field doesn't end on a char or int boundary.  If it does
-            ** end on a boundary, then zeros have already been shifted in.  g_and emits no code
-            ** if the mask is all ones.
-            */
-            if (Expr->BitOffs + Expr->BitWidth != CHAR_BITS &&
-                Expr->BitOffs + Expr->BitWidth != INT_BITS) {
-                g_and (F, (0x0001U << Expr->BitWidth) - 1U);
+            if (ED_NeedsTest (Expr)) {
+                /* If we need to do a test, then we avoid shifting (ASR only shifts one bit
+                ** at a time, so is slow) and just AND with the appropriate mask, then test
+                ** the result of that.
+                */
+
+                /* Avoid overly large shift. */
+                if (EndBit == sizeof (unsigned long) * CHAR_BIT) {
```

Comment:
For the sake of consistency, perhaps `CHAR_BITS` (from "src/cc65/datatype.h") should be used here instead of `CHAR_BIT` (from <limits.h>), and maybe `SizeOf (type_ulong)` insteadof `sizeof (unsigned long)`.

Or maybe `LONG_BITS` (well, `ULONG_BITS` is missing, so this alternative is weak) to replace the whole expression on the right side of the comparison.

Reply:
Oh, I see it. the point here is to avoid over-shifting on the compiler side.

> 
> 
> For the sake of consistency, perhaps `CHAR_BITS` (from "src/cc65/datatype.h") should be used here instead of `CHAR_BIT` (from <limits.h>), and maybe `SizeOf (type_ulong)` insteadof `sizeof (unsigned long)`.
> 
> Or maybe `LONG_BITS` (well, `ULONG_BITS` is missing, so this alternative is weak) to replace the whole expression on the right side of the comparison.

Oh, I see it. the point here is to avoid over-shifting on the compiler side.

Reply:
> It is annoying that we didn't use int32_t/uint32_t instead of long/unsigned long

I updated the comment.

> Or maybe LONG_BIT

That's just POSIX, so somewhat less standard.

> It is annoying that we didn't use int32_t/uint32_t instead of long/unsigned long

They probably didn't exist when the code was written. @oliverschmidt would you merge a PR that did `s/unsigned long/uint32_t` (or `uint_fast32_t`), etc?

Reply:
> They probably didn't exist when the code was written. @oliverschmidt would you merge a PR that did `s/unsigned long/uint32_t` (or `uint_fast32_t`), etc?

There was a proposal to deal with 64-bit compilation warnings (#1009) which I think should be a good chance to switch to stdint. Still... it could affect every single file in the code base. I have 200+ unmerged commits that could be impacted. XD

Reply:
>> It is annoying that we didn't use int32_t/uint32_t instead of long/unsigned long.
>
>They probably didn't exist when the code was written.

Uz was proud of the fact that he didn't use any C99 code in the tools.  He bragged that we could use any ANSI C compiler to build those tools.  Neither `<stdint.h>` nor `<stdbool.h>` are used anywhere.

Reply:
Wow really? Ever since we started using those things in VICE we keep finding new subtle bugs related to surprise compiling due to "long" and "int"

Reply:
I forgot to mention that Uz went overboard:  he even didn't use `NULL`!  It frustrates the heck out of me.  Every time that I see a zero being assigned to something, I wonder, "is that an integer or a pointer??"

Reply:
> Uz was proud of the fact that he didn't use any C99 code in the tools.

That was probably reasonable until 2010 or so. It's 2020 now, so C99 if not C11 can be relied on.